File: solutions/ocaml/binary-search-tree/1/binary_search_tree.ml
open Base

type bst = Empty | Node of int * bst * bst

let empty = Empty

let value = function
  | Empty -> Error "Empty tree has no value"
  | Node (v, _, _) -> Ok v

let left = function
  | Empty -> Error "Empty tree has no left child"
  | Node (_, l, _) -> Ok l

let right = function
  | Empty -> Error "Empty tree has no right child"
  | Node (_, _, r) -> Ok r

let insert v t =
  let rec inserter = function
    | Empty -> Node (v, Empty, Empty)
    | Node (node_val, left_tree, right_tree) ->
        if v <= node_val then Node (node_val, inserter left_tree, right_tree)
        else Node (node_val, left_tree, inserter right_tree)
  in
  inserter t

let to_list t =
  let rec traverse = function
    | Empty -> []
    | Node (v, l, r) -> traverse l @ [ v ] @ traverse r
  in
  traverse t

File: solutions/elixir/binary-search-tree/1/lib/binary_search_tree.ex
defmodule BinarySearchTree do
  @type bst_node :: %{data: any, left: bst_node | nil, right: bst_node | nil}

  @doc """
  Create a new Binary Search Tree with root's value as the given 'data'
  """
  @spec new(any) :: bst_node
  def new(data) do
    %{data: data, left: nil, right: nil}
  end

  @doc """
  Creates and inserts a node with its value as 'data' into the tree.
  """
  @spec insert(bst_node, any) :: bst_node
  def insert(nil, data), do: new(data)

  def insert(%{data: node_data, left: left_tree, right: right_tree}, data) do
    cond do
      data <= node_data -> %{data: node_data, left: insert(left_tree, data), right: right_tree}
      true -> %{data: node_data, left: left_tree, right: insert(right_tree, data)}
    end
  end

  @doc """
  Traverses the Binary Search Tree in order and returns a list of each node's data.
  """
  @spec in_order(bst_node) :: [any]
  def in_order(nil), do: []

  def in_order(%{data: d, left: l, right: r}) do
    in_order(l) ++ [d] ++ in_order(r)
  end
end
